calyx/opt/src/passes_experimental/metadata_table_gen.rs

```diff
@@ -55,57 +75,168 @@ impl Named for Metadata {
+    //     if ctx.metadata.is_none() {
+    //         let mut table = Metadata::new();
+    //         for component in &ctx.components {
+    //             let cmpt_iter = component.groups.into_iter();
+    //             for rcc_grp in cmpt_iter {
+    //                 let grp = rcc_grp.borrow_mut();
+    //                 let pos_data = grp.attributes.copy_span();
+    //                 let (file, span) = pos_data.get_line_num();
+    //                 table.add_entry(
+    //                     component.name,
+    //                     grp.name(),
+    //                     span,
+    //                     PathBuf::from(file),
+    //                 );
+    //                 // hm may need to actually use the full name of the group
+    //             }
+
+    //             ctx.metadata = Some(table.to_string());
+    //         }
+    //         // to generate the table:
+    //         let mut new_src_table: SourceInfoTable =
+    //             match &ctx.source_info_table {
+    //                 None => SourceInfoTable::new_empty(),
+    //                 Some(s) => s.clone(),
+    //             };
+
+    //         // add this file to the table
+    //         // longass train of calls that inlines what i do above to get file name from component span
+    //         let temp0 = &ctx.components[0]
+    //             .groups
+    //             .into_iter()
+    //             .next()
+    //             .unwrap()
+    //             .borrow_mut()
+    //             .attributes
+    //             .copy_span();
+    //         let (file, _span) = temp0.get_line_num();
+    //         let id = new_src_table.push_file(PathBuf::from(file));
+    //         self.src_table = new_src_table;
+    //         self.file_ids.insert(String::from(file), id);
+    //         //update this -> table needs its own to_string
+    //         // may need to add display method to source info table (source_info.rs)
+    //     }
+    //     Ok(Action::Stop)
+    // }
+
+    // this visits each component
+    fn start(
+        &mut self,
+        comp: &mut calyx_ir::Component,
+        _sigs: &calyx_ir::LibrarySignatures,
+        _comps: &[calyx_ir::Component],
+    ) -> VisResult {
+        // implement visiting each component and adding the positions
+        // how this works:
+        // components have attributes !!
+        // check if file exists for component definition
+        let binding = comp.attributes.copy_span();
+        let (file, _bs) = binding.get_line_num();
+
+        // add file to source table (if not already in)
+        if !self.file_ids.contains_key(file) {
+            let id = self.src_table.push_file(PathBuf::from(file));
+            self.file_ids.insert(String::from(file), id);
+            //dbg!(&self.src_table);
         }
-        Ok(Action::Stop)
+        // visit all groups in component
+        comp.groups.iter().for_each(|rrcgrp| {
```

Comment:
This is a bit of a stylistic nit, so others might disagree. But I think a `for` loop would be clearer
here than `for_each` and I don't tend to see a good argument for the latter when
the loop itself has side-effects

Reply:
The end result is the same regardless